File: apple-silicon-monitoring/llm-telemetry/python/llm_telemetry/telemetry.py
"""
Main telemetry module for LLM observability
"""

import time
import logging
from typing import Optional, List, Dict, Any
from dataclasses import dataclass
from contextlib import contextmanager

from opentelemetry import metrics
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_VERSION

from .detection import (
    compute_entropy,
    compute_repetition_score,
    compute_perplexity,
    detect_refusal,
    compute_confidence_stats,
)
from .metrics import ErrorType

logger = logging.getLogger(__name__)

# Global meter provider
_meter_provider: Optional[MeterProvider] = None


def init_telemetry(
    endpoint: str = "localhost:4317",
    service_name: str = "llm-inference",
    service_version: str = "1.0.0",
    export_interval_ms: int = 10000,
    insecure: bool = True,
    headers: Optional[Dict[str, str]] = None,
) -> MeterProvider:
    """
    Initialize OpenTelemetry metrics with OTLP exporter.
    
    Args:
        endpoint: OTLP collector endpoint
        service_name: Name of the service
        service_version: Version of the service
        export_interval_ms: How often to export metrics
        insecure: Whether to use insecure connection
        headers: Optional headers for authentication
    
    Returns:
        Configured MeterProvider
    """
    global _meter_provider
    
    if _meter_provider is not None:
        logger.warning("Telemetry already initialized, returning existing provider")
        return _meter_provider
    
    # Create resource with service info
    resource = Resource.create({
        SERVICE_NAME: service_name,
        SERVICE_VERSION: service_version,
        "platform": "apple-silicon",
    })
    
    # Create OTLP exporter
    exporter = OTLPMetricExporter(
        endpoint=endpoint,
        insecure=insecure,
        headers=headers or {},
    )
    
    # Create metric reader
    reader = PeriodicExportingMetricReader(
        exporter,
        export_interval_millis=export_interval_ms,
    )
    
    # Create and set meter provider
    _meter_provider = MeterProvider(
        resource=resource,
        metric_readers=[reader],
    )
    metrics.set_meter_provider(_meter_provider)
    
    logger.info(f"Initialized LLM telemetry, exporting to {endpoint}")
    return _meter_provider


@dataclass
class InferenceResult:
    """Result of an inference operation"""
    prompt_tokens: int
    output_tokens: int
    prompt_latency_s: float
    inference_latency_s: float
    token_probs: Optional[List[float]] = None
    output_text: Optional[str] = None
    error: Optional[ErrorType] = None
    context_length: Optional[int] = None
    batch_size: int = 1


class LLMTelemetry:
    """
    Telemetry collector for LLM inference workloads.
    
    Provides metrics for:
    - Request throughput and latency
    - Token processing statistics
    - GPU memory allocation
    - Hallucination detection signals
    - Error classification
    - Model health scoring
    """
    
    def __init__(
        self,
        model_name: str,
        model_version: str = "unknown",
        gpu_id: str = "0",
    ):
        """
        Initialize telemetry for a specific model.
        
        Args:
            model_name: Name of the model (e.g., "llama-3-70b")
            model_version: Version of the model
            gpu_id: GPU identifier
        """
        self.model_name = model_name
        self.model_version = model_version
        self.gpu_id = gpu_id
        self.labels = {
            "model": model_name,
            "model_version": model_version,
            "gpu": gpu_id,
        }
        
        # Get meter
        self.meter = metrics.get_meter("llm-inference", "1.0.0")
        
        # Initialize metrics
        self._init_metrics()
        
        # State tracking
        self._request_count = 0
        self._error_count = 0
        self._last_entropy_values: List[float] = []
        
    def _init_metrics(self):
        """Initialize all metric instruments"""
        
        # Request metrics
        self.inference_requests = self.meter.create_counter(
            "llm_inference_requests_total",
            description="Total number of inference requests",
            unit="1",
        )
        
        self.inference_duration = self.meter.create_histogram(
            "llm_inference_duration_seconds",
            description="Inference latency by phase (prompt/generation)",
            unit="s",
        )
        
        # Token metrics
        self.tokens_processed = self.meter.create_counter(
            "llm_tokens_processed_total",
            description="Total tokens processed (input/output)",
            unit="1",
        )
        
        self.tokens_per_second = self.meter.create_gauge(
            "llm_tokens_per_second",
            description="Current token generation throughput",
            unit="1/s",
        )
        
        # Batch and context
        self.batch_size_hist = self.meter.create_histogram(
            "llm_batch_size",
            description="Batch sizes processed",
            unit="1",
        )
        
        self.context_length_hist = self.meter.create_histogram(
            "llm_context_length",
            description="Context window utilization",
            unit="1",
        )
        
        # GPU memory
        self.gpu_memory_allocated = self.meter.create_gauge(
            "llm_gpu_memory_allocated_bytes",
            description="GPU memory allocated for model",
            unit="By",
        )
        
        # KV cache
        self.kv_cache_utilization = self.meter.create_gauge(
            "llm_kv_cache_utilization",
            description="KV cache utilization (0-1)",
            unit="1",
        )
        
        # Queue depth
        self.queue_depth = self.meter.create_gauge(
            "llm_queue_depth",
            description="Number of pending requests",
            unit="1",
        )
        
        # Errors
        self.errors_total = self.meter.create_counter(
            "llm_error_total",
            description="Total errors by type",
            unit="1",
        )
        
        # Model status
        self.model_loaded = self.meter.create_gauge(
            "llm_model_loaded",
            description="Whether the model is loaded and ready",
            unit="1",
        )
        
        # Hallucination detection metrics
        self.output_entropy = self.meter.create_histogram(
            "llm_output_entropy",
            description="Token probability entropy (higher = more uncertain)",
            unit="1",
        )
        
        self.confidence_mean = self.meter.create_gauge(
            "llm_confidence_mean",
            description="Mean token confidence score",
            unit="1",
        )
        
        self.confidence_std = self.meter.create_gauge(
            "llm_confidence_std",
            description="Standard deviation of token confidence",
            unit="1",
        )
        
        self.repetition_score = self.meter.create_gauge(
            "llm_repetition_score",
            description="N-gram repetition score (0-1)",
            unit="1",
        )
        
        self.perplexity = self.meter.create_histogram(
            "llm_perplexity",
            description="Output perplexity",
            unit="1",
        )
        
        self.refusal_rate = self.meter.create_gauge(
            "llm_refusal_rate",
            description="Rate of refusal responses",
            unit="1",
        )
        
        # Health score
        self.model_health_score = self.meter.create_gauge(
            "llm_model_health_score",
            description="Composite model health score (0-100)",
            unit="1",
        )
    
    def record_inference(
        self,
        prompt_tokens: int,
        output_tokens: int,
        prompt_latency_s: float,
        inference_latency_s: float,
        token_probs: Optional[List[float]] = None,
        output_text: Optional[str] = None,
        error: Optional[ErrorType] = None,
        context_length: Optional[int] = None,
        batch_size: int = 1,
    ):
        """
        Record metrics for a completed inference.
        
        Args:
            prompt_tokens: Number of tokens in the prompt
            output_tokens: Number of generated tokens
            prompt_latency_s: Time to process prompt (prefill)
            inference_latency_s: Time to generate output (decode)
            token_probs: Optional list of token probabilities for hallucination detection
            output_text: Optional generated text for analysis
            error: Optional error type if inference failed
            context_length: Optional total context length
            batch_size: Number of requests in batch
        """
        base_labels = self.labels.copy()
        
        # Request count
        status = "error" if error else "success"
        self.inference_requests.add(1, {**base_labels, "status": status})
        self._request_count += 1
        
        # Error handling
        if error:
            self.errors_total.add(1, {**base_labels, "error_type": error.value})
            self._error_count += 1
            return
        
        # Latency metrics
        self.inference_duration.record(prompt_latency_s, {**base_labels, "phase": "prompt"})
        self.inference_duration.record(inference_latency_s, {**base_labels, "phase": "generation"})
        
        total_latency = prompt_latency_s + inference_latency_s
        self.inference_duration.record(total_latency, {**base_labels, "phase": "total"})
        
        # Token metrics
        self.tokens_processed.add(prompt_tokens, {**base_labels, "direction": "input"})
        self.tokens_processed.add(output_tokens, {**base_labels, "direction": "output"})
        
        # Throughput
        if inference_latency_s > 0:
            tps = output_tokens / inference_latency_s
            self.tokens_per_second.set(tps, base_labels)
        
        # Batch and context
        self.batch_size_hist.record(batch_size, base_labels)
        if context_length:
            self.context_length_hist.record(context_length, base_labels)
        
        # Hallucination detection
        if token_probs:
            self._record_hallucination_metrics(token_probs, output_text, base_labels)
    
    def _record_hallucination_metrics(
        self,
        token_probs: List[float],
        output_text: Optional[str],
        labels: Dict[str, str],
    ):
        """Record hallucination detection metrics"""
        
        # Entropy
        entropy = compute_entropy(token_probs)
        self.output_entropy.record(entropy, labels)
        self._last_entropy_values.append(entropy)
        if len(self._last_entropy_values) > 100:
            self._last_entropy_values.pop(0)
        
        # Confidence statistics
        mean_conf, std_conf = compute_confidence_stats(token_probs)
        self.confidence_mean.set(mean_conf, labels)
        self.confidence_std.set(std_conf, labels)
        
        # Perplexity
        ppl = compute_perplexity(token_probs)
        self.perplexity.record(ppl, labels)
        
        # Text-based metrics
        if output_text:
            # Repetition
            rep_score = compute_repetition_score(output_text, n=3)
            self.repetition_score.set(rep_score, labels)
            
            # Refusal detection
            is_refusal = detect_refusal(output_text)
            # Update running refusal rate
            # This is simplified; in production use a proper windowed rate
            self.refusal_rate.set(1.0 if is_refusal else 0.0, labels)
    
    def record_gpu_memory(self, allocated_bytes: int):
        """Record current GPU memory allocation"""
        self.gpu_memory_allocated.set(allocated_bytes, self.labels)
    
    def record_kv_cache(self, utilization: float):
        """Record KV cache utilization (0-1)"""
        self.kv_cache_utilization.set(utilization, self.labels)
    
    def record_queue_depth(self, depth: int):
        """Record current queue depth"""
        self.queue_depth.set(depth, self.labels)
    
    def set_model_loaded(self, loaded: bool):
        """Set model loaded status"""
        self.model_loaded.set(1.0 if loaded else 0.0, self.labels)
    
    def compute_health_score(self) -> float:
        """
        Compute a composite health score (0-100) based on recent metrics.
        
        Factors:
        - Error rate (lower is better)
        - Average entropy (lower is better)
        - Repetition score (lower is better)
        
        Returns:
            Health score from 0 to 100
        """
        score = 100.0
        
        # Error rate penalty (up to 50 points)
        if self._request_count > 0:
            error_rate = self._error_count / self._request_count
            score -= min(50, error_rate * 100)
        
        # Entropy penalty (up to 25 points)
        if self._last_entropy_values:
            avg_entropy = sum(self._last_entropy_values) / len(self._last_entropy_values)
            # Normalize: entropy > 4 is concerning
            entropy_penalty = min(25, (avg_entropy / 4) * 25)
            score -= entropy_penalty
        
        # Ensure score is in valid range
        score = max(0, min(100, score))
        
        # Record the score
        self.model_health_score.set(score, self.labels)
        
        return score
    
    @contextmanager
    def inference_span(self, prompt_tokens: int, batch_size: int = 1):
        """
        Context manager for timing an inference operation.
        
        Usage:
            with telemetry.inference_span(prompt_tokens=100) as span:
                # Prompt processing
                span.mark_prompt_done()
                # Generation
                output = generate(...)
                span.set_output(output_tokens=50, token_probs=[...], text="...")
        """
        span = InferenceSpan(self, prompt_tokens, batch_size)
        try:
            yield span
        except Exception as e:
            span.set_error(ErrorType.from_exception(e))
            raise
        finally:
            span.finish()


class InferenceSpan:
    """Helper class for timing inference operations"""
    
    def __init__(self, telemetry: LLMTelemetry, prompt_tokens: int, batch_size: int):
        self.telemetry = telemetry
        self.prompt_tokens = prompt_tokens
        self.batch_size = batch_size
        self.start_time = time.perf_counter()
        self.prompt_end_time: Optional[float] = None
        self.output_tokens: int = 0
        self.token_probs: Optional[List[float]] = None
        self.output_text: Optional[str] = None
        self.error: Optional[ErrorType] = None
        self.context_length: Optional[int] = None
    
    def mark_prompt_done(self):
        """Mark the end of prompt processing"""
        self.prompt_end_time = time.perf_counter()
    
    def set_output(
        self,
        output_tokens: int,
        token_probs: Optional[List[float]] = None,
        text: Optional[str] = None,
        context_length: Optional[int] = None,
    ):
        """Set output information"""
        self.output_tokens = output_tokens
        self.token_probs = token_probs
        self.output_text = text
        self.context_length = context_length
    
    def set_error(self, error: ErrorType):
        """Set error information"""
        self.error = error
    
    def finish(self):
        """Finish the span and record metrics"""
        end_time = time.perf_counter()
        
        if self.prompt_end_time:
            prompt_latency = self.prompt_end_time - self.start_time
            inference_latency = end_time - self.prompt_end_time
        else:
            # No prompt marker, split time proportionally
            total_time = end_time - self.start_time
            prompt_latency = total_time * 0.1  # Estimate
            inference_latency = total_time * 0.9
        
        self.telemetry.record_inference(
            prompt_tokens=self.prompt_tokens,
            output_tokens=self.output_tokens,
            prompt_latency_s=prompt_latency,
            inference_latency_s=inference_latency,
            token_probs=self.token_probs,
            output_text=self.output_text,
            error=self.error,
            context_length=self.context_length,
            batch_size=self.batch_size,
        )

File: apple-silicon-monitoring/llm-telemetry/go/telemetry.go
// Package llmtelemetry provides OpenTelemetry instrumentation for LLM inference workloads.
//
// Features:
//   - Inference latency and throughput metrics
//   - Token processing statistics
//   - GPU memory tracking
//   - Hallucination detection signals (entropy, repetition, confidence)
//   - Error classification and model health scoring
//
// Usage:
//
//	telemetry, _ := llmtelemetry.New("llama-3-70b",
//	    llmtelemetry.WithEndpoint("localhost:4317"),
//	)
//	defer telemetry.Shutdown()
//
//	telemetry.RecordInference(llmtelemetry.InferenceResult{
//	    PromptTokens:     100,
//	    OutputTokens:     50,
//	    PromptLatency:    500 * time.Millisecond,
//	    InferenceLatency: 2 * time.Second,
//	})
package llmtelemetry

import (
	"context"
	"math"
	"strings"
	"sync"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
	"go.opentelemetry.io/otel/metric"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
)

// ErrorType classifies inference errors
type ErrorType string

const (
	ErrorInvalidInput    ErrorType = "invalid_input"
	ErrorContextTooLong  ErrorType = "context_too_long"
	ErrorOutOfMemory     ErrorType = "out_of_memory"
	ErrorTimeout         ErrorType = "timeout"
	ErrorKernelFailure   ErrorType = "kernel_failure"
	ErrorMetalCrash      ErrorType = "metal_crash"
	ErrorModelNotLoaded  ErrorType = "model_not_loaded"
	ErrorUnknown         ErrorType = "unknown"
)

// InferenceResult contains the results of an inference operation
type InferenceResult struct {
	PromptTokens     int
	OutputTokens     int
	PromptLatency    time.Duration
	InferenceLatency time.Duration
	TokenProbs       []float64 // Optional: token probabilities for hallucination detection
	OutputText       string    // Optional: generated text for analysis
	Error            ErrorType // Empty if no error
	ContextLength    int       // Optional: total context length
	BatchSize        int       // Default: 1
}

// Options configures the telemetry instance
type Options struct {
	Endpoint       string
	ServiceName    string
	ServiceVersion string
	Insecure       bool
	ExportInterval time.Duration
	ModelVersion   string
	GPUID          string
}

// Option is a functional option for configuring telemetry
type Option func(*Options)

// WithEndpoint sets the OTLP endpoint
func WithEndpoint(endpoint string) Option {
	return func(o *Options) {
		o.Endpoint = endpoint
	}
}

// WithServiceName sets the service name
func WithServiceName(name string) Option {
	return func(o *Options) {
		o.ServiceName = name
	}
}

// WithInsecure enables insecure connection
func WithInsecure(insecure bool) Option {
	return func(o *Options) {
		o.Insecure = insecure
	}
}

// WithModelVersion sets the model version
func WithModelVersion(version string) Option {
	return func(o *Options) {
		o.ModelVersion = version
	}
}

// WithGPUID sets the GPU identifier
func WithGPUID(id string) Option {
	return func(o *Options) {
		o.GPUID = id
	}
}

// LLMTelemetry provides metrics collection for LLM inference
type LLMTelemetry struct {
	modelName    string
	modelVersion string
	gpuID        string
	
	meter        metric.Meter
	provider     *sdkmetric.MeterProvider
	
	// Metrics
	inferenceRequests    metric.Int64Counter
	inferenceDuration    metric.Float64Histogram
	tokensProcessed      metric.Int64Counter
	tokensPerSecond      metric.Float64Gauge
	batchSize            metric.Int64Histogram
	contextLength        metric.Int64Histogram
	gpuMemoryAllocated   metric.Int64Gauge
	kvCacheUtilization   metric.Float64Gauge
	queueDepth           metric.Int64Gauge
	errorsTotal          metric.Int64Counter
	modelLoaded          metric.Int64Gauge
	outputEntropy        metric.Float64Histogram
	confidenceMean       metric.Float64Gauge
	repetitionScore      metric.Float64Gauge
	perplexity           metric.Float64Histogram
	modelHealthScore     metric.Float64Gauge
	
	// State tracking
	mu              sync.Mutex
	requestCount    int64
	errorCount      int64
	entropyHistory  []float64
	attrs           []attribute.KeyValue
}

// New creates a new LLMTelemetry instance
func New(modelName string, opts ...Option) (*LLMTelemetry, error) {
	options := &Options{
		Endpoint:       "localhost:4317",
		ServiceName:    "llm-inference",
		ServiceVersion: "1.0.0",
		Insecure:       true,
		ExportInterval: 10 * time.Second,
		ModelVersion:   "unknown",
		GPUID:          "0",
	}
	
	for _, opt := range opts {
		opt(options)
	}
	
	ctx := context.Background()
	
	// Create OTLP exporter
	exporterOpts := []otlpmetricgrpc.Option{
		otlpmetricgrpc.WithEndpoint(options.Endpoint),
	}
	if options.Insecure {
		exporterOpts = append(exporterOpts, otlpmetricgrpc.WithInsecure())
	}
	
	exporter, err := otlpmetricgrpc.New(ctx, exporterOpts...)
	if err != nil {
		return nil, err
	}
	
	// Create resource
	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName(options.ServiceName),
			semconv.ServiceVersion(options.ServiceVersion),
			attribute.String("platform", "apple-silicon"),
		),
	)
	if err != nil {
		return nil, err
	}
	
	// Create meter provider
	provider := sdkmetric.NewMeterProvider(
		sdkmetric.WithResource(res),
		sdkmetric.WithReader(
			sdkmetric.NewPeriodicReader(exporter,
				sdkmetric.WithInterval(options.ExportInterval),
			),
		),
	)
	
	otel.SetMeterProvider(provider)
	
	t := &LLMTelemetry{
		modelName:    modelName,
		modelVersion: options.ModelVersion,
		gpuID:        options.GPUID,
		provider:     provider,
		meter:        provider.Meter("llm-inference"),
		attrs: []attribute.KeyValue{
			attribute.String("model", modelName),
			attribute.String("model_version", options.ModelVersion),
			attribute.String("gpu", options.GPUID),
		},
		entropyHistory: make([]float64, 0, 100),
	}
	
	if err := t.initMetrics(); err != nil {
		return nil, err
	}
	
	return t, nil
}

func (t *LLMTelemetry) initMetrics() error {
	var err error
	
	t.inferenceRequests, err = t.meter.Int64Counter("llm_inference_requests_total",
		metric.WithDescription("Total number of inference requests"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return err
	}
	
	t.inferenceDuration, err = t.meter.Float64Histogram("llm_inference_duration_seconds",
		metric.WithDescription("Inference latency by phase"),
		metric.WithUnit("s"),
	)
	if err != nil {
		return err
	}
	
	t.tokensProcessed, err = t.meter.Int64Counter("llm_tokens_processed_total",
		metric.WithDescription("Total tokens processed"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return err
	}
	
	t.tokensPerSecond, err = t.meter.Float64Gauge("llm_tokens_per_second",
		metric.WithDescription("Current token generation throughput"),
		metric.WithUnit("1/s"),
	)
	if err != nil {
		return err
	}
	
	t.batchSize, err = t.meter.Int64Histogram("llm_batch_size",
		metric.WithDescription("Batch sizes processed"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return err
	}
	
	t.contextLength, err = t.meter.Int64Histogram("llm_context_length",
		metric.WithDescription("Context window utilization"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return err
	}
	
	t.gpuMemoryAllocated, err = t.meter.Int64Gauge("llm_gpu_memory_allocated_bytes",
		metric.WithDescription("GPU memory allocated for model"),
		metric.WithUnit("By"),
	)
	if err != nil {
		return err
	}
	
	t.kvCacheUtilization, err = t.meter.Float64Gauge("llm_kv_cache_utilization",
		metric.WithDescription("KV cache utilization"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return err
	}
	
	t.queueDepth, err = t.meter.Int64Gauge("llm_queue_depth",
		metric.WithDescription("Number of pending requests"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return err
	}
	
	t.errorsTotal, err = t.meter.Int64Counter("llm_error_total",
		metric.WithDescription("Total errors by type"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return err
	}
	
	t.modelLoaded, err = t.meter.Int64Gauge("llm_model_loaded",
		metric.WithDescription("Whether the model is loaded"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return err
	}
	
	t.outputEntropy, err = t.meter.Float64Histogram("llm_output_entropy",
		metric.WithDescription("Token probability entropy"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return err
	}
	
	t.confidenceMean, err = t.meter.Float64Gauge("llm_confidence_mean",
		metric.WithDescription("Mean token confidence score"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return err
	}
	
	t.repetitionScore, err = t.meter.Float64Gauge("llm_repetition_score",
		metric.WithDescription("N-gram repetition score"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return err
	}
	
	t.perplexity, err = t.meter.Float64Histogram("llm_perplexity",
		metric.WithDescription("Output perplexity"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return err
	}
	
	t.modelHealthScore, err = t.meter.Float64Gauge("llm_model_health_score",
		metric.WithDescription("Composite model health score"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return err
	}
	
	return nil
}

// RecordInference records metrics for a completed inference
func (t *LLMTelemetry) RecordInference(ctx context.Context, result InferenceResult) {
	t.mu.Lock()
	defer t.mu.Unlock()
	
	batchSize := result.BatchSize
	if batchSize < 1 {
		batchSize = 1
	}
	
	// Request count
	status := "success"
	if result.Error != "" {
		status = "error"
	}
	
	attrs := append(t.attrs, attribute.String("status", status))
	t.inferenceRequests.Add(ctx, 1, metric.WithAttributes(attrs...))
	t.requestCount++
	
	// Handle errors
	if result.Error != "" {
		errorAttrs := append(t.attrs, attribute.String("error_type", string(result.Error)))
		t.errorsTotal.Add(ctx, 1, metric.WithAttributes(errorAttrs...))
		t.errorCount++
		return
	}
	
	// Latency
	promptAttrs := append(t.attrs, attribute.String("phase", "prompt"))
	genAttrs := append(t.attrs, attribute.String("phase", "generation"))
	totalAttrs := append(t.attrs, attribute.String("phase", "total"))
	
	t.inferenceDuration.Record(ctx, result.PromptLatency.Seconds(), metric.WithAttributes(promptAttrs...))
	t.inferenceDuration.Record(ctx, result.InferenceLatency.Seconds(), metric.WithAttributes(genAttrs...))
	t.inferenceDuration.Record(ctx, (result.PromptLatency + result.InferenceLatency).Seconds(), metric.WithAttributes(totalAttrs...))
	
	// Tokens
	inputAttrs := append(t.attrs, attribute.String("direction", "input"))
	outputAttrs := append(t.attrs, attribute.String("direction", "output"))
	
	t.tokensProcessed.Add(ctx, int64(result.PromptTokens), metric.WithAttributes(inputAttrs...))
	t.tokensProcessed.Add(ctx, int64(result.OutputTokens), metric.WithAttributes(outputAttrs...))
	
	// Throughput
	if result.InferenceLatency > 0 {
		tps := float64(result.OutputTokens) / result.InferenceLatency.Seconds()
		t.tokensPerSecond.Record(ctx, tps, metric.WithAttributes(t.attrs...))
	}
	
	// Batch and context
	t.batchSize.Record(ctx, int64(batchSize), metric.WithAttributes(t.attrs...))
	if result.ContextLength > 0 {
		t.contextLength.Record(ctx, int64(result.ContextLength), metric.WithAttributes(t.attrs...))
	}
	
	// Hallucination detection
	if len(result.TokenProbs) > 0 {
		t.recordHallucinationMetrics(ctx, result.TokenProbs, result.OutputText)
	}
}

func (t *LLMTelemetry) recordHallucinationMetrics(ctx context.Context, probs []float64, text string) {
	// Entropy
	entropy := computeEntropy(probs)
	t.outputEntropy.Record(ctx, entropy, metric.WithAttributes(t.attrs...))
	
	t.entropyHistory = append(t.entropyHistory, entropy)
	if len(t.entropyHistory) > 100 {
		t.entropyHistory = t.entropyHistory[1:]
	}
	
	// Confidence
	mean := computeMean(probs)
	t.confidenceMean.Record(ctx, mean, metric.WithAttributes(t.attrs...))
	
	// Perplexity
	ppl := computePerplexity(probs)
	t.perplexity.Record(ctx, ppl, metric.WithAttributes(t.attrs...))
	
	// Repetition (if text provided)
	if text != "" {
		rep := computeRepetition(text)
		t.repetitionScore.Record(ctx, rep, metric.WithAttributes(t.attrs...))
	}
}

// RecordGPUMemory records current GPU memory allocation
func (t *LLMTelemetry) RecordGPUMemory(ctx context.Context, bytes int64) {
	t.gpuMemoryAllocated.Record(ctx, bytes, metric.WithAttributes(t.attrs...))
}

// RecordKVCache records KV cache utilization
func (t *LLMTelemetry) RecordKVCache(ctx context.Context, utilization float64) {
	t.kvCacheUtilization.Record(ctx, utilization, metric.WithAttributes(t.attrs...))
}

// RecordQueueDepth records current queue depth
func (t *LLMTelemetry) RecordQueueDepth(ctx context.Context, depth int64) {
	t.queueDepth.Record(ctx, depth, metric.WithAttributes(t.attrs...))
}

// SetModelLoaded sets the model loaded status
func (t *LLMTelemetry) SetModelLoaded(ctx context.Context, loaded bool) {
	val := int64(0)
	if loaded {
		val = 1
	}
	t.modelLoaded.Record(ctx, val, metric.WithAttributes(t.attrs...))
}

// ComputeHealthScore computes and records the model health score
func (t *LLMTelemetry) ComputeHealthScore(ctx context.Context) float64 {
	t.mu.Lock()
	defer t.mu.Unlock()
	
	score := 100.0
	
	// Error rate penalty
	if t.requestCount > 0 {
		errorRate := float64(t.errorCount) / float64(t.requestCount)
		score -= math.Min(50, errorRate*100)
	}
	
	// Entropy penalty
	if len(t.entropyHistory) > 0 {
		avgEntropy := computeMean(t.entropyHistory)
		entropyPenalty := math.Min(25, (avgEntropy/4)*25)
		score -= entropyPenalty
	}
	
	score = math.Max(0, math.Min(100, score))
	t.modelHealthScore.Record(ctx, score, metric.WithAttributes(t.attrs...))
	
	return score
}

// Shutdown shuts down the telemetry provider
func (t *LLMTelemetry) Shutdown(ctx context.Context) error {
	return t.provider.Shutdown(ctx)
}

// Helper functions

func computeEntropy(probs []float64) float64 {
	if len(probs) == 0 {
		return 0
	}
	
	entropy := 0.0
	for _, p := range probs {
		if p > 1e-10 {
			entropy -= p * math.Log(p+1e-10)
		}
	}
	return entropy
}

func computePerplexity(probs []float64) float64 {
	if len(probs) == 0 {
		return 1
	}
	
	logProbSum := 0.0
	for _, p := range probs {
		logProbSum += math.Log(p + 1e-10)
	}
	
	return math.Exp(-logProbSum / float64(len(probs)))
}

func computeMean(vals []float64) float64 {
	if len(vals) == 0 {
		return 0
	}
	
	sum := 0.0
	for _, v := range vals {
		sum += v
	}
	return sum / float64(len(vals))
}

func computeRepetition(text string) float64 {
	words := strings.Fields(strings.ToLower(text))
	n := 3
	
	if len(words) < n {
		return 0
	}
	
	ngrams := make(map[string]int)
	for i := 0; i <= len(words)-n; i++ {
		key := strings.Join(words[i:i+n], " ")
		ngrams[key]++
	}
	
	repeated := 0
	for _, count := range ngrams {
		if count > 1 {
			repeated += count - 1
		}
	}
	
	total := len(words) - n + 1
	if total <= 0 {
		return 0
	}
	
	return float64(repeated) / float64(total)
}
